server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -69,15 +149,51 @@ public void fetchObject(
                 .onError(responseObserver)
                 .submit(() -> {
                     final Object o = object.get();
-                    final FetchObjectResponse response = serialize(type, session, o);
-                    responseObserver.onNext(response);
-                    responseObserver.onCompleted();
+                    ObjectType objectTypeInstance = getObjectTypeInstance(type, o);
+
+                    StreamObserver<StreamResponse> wrappedResponseObserver = new StreamObserver<>() {
+                        @Override
+                        public void onNext(StreamResponse value) {
+                            // TODO Do we send a message here, _then_ error if the plugin didn't close? Or do we close
+                            // with an error without sending anything (and release exports)?
+                            responseObserver.onNext(FetchObjectResponse.newBuilder()
+                                    .setType(type)
+                                    .setData(value.getData().getPayload())
+                                    .addAllTypedExportIds(value.getData().getTypedExportIdsList())
+                                    .build());
+                        }
+
+                        @Override
+                        public void onError(Throwable t) {
+                            responseObserver.onError(t);
+                        }
+
+                        @Override
+                        public void onCompleted() {
+                            responseObserver.onCompleted();
+                        }
+                    };
+                    PluginMessageSender connection = new PluginMessageSender(wrappedResponseObserver, session);
+                    objectTypeInstance.clientConnection(o, connection);
+
+                    if (!connection.isClosed()) {
+                        connection.onClose();
+                        throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT,
```

Comment:
We talked about having a recording response observer that records the plugin's actions, and allows either of:
1. One response + one close
2. One error

Otherwise, we send a "bad plugin" error.

Reply:
I think this is resolved now (PR comments are seemingly hiding at random, I know there are two other threads around this, I'll update those too as I find them...)